dnscrypt-proxy/odoh.go

```diff
@@ -0,0 +1,185 @@
+package main
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"encoding/binary"
+	"fmt"
+
+	hpkecompact "github.com/jedisct1/go-hpke-compact"
+)
+
+const (
+	odohVersion = uint16(0xff06)
+)
+
+type ODoHTarget struct {
+	suite     *hpkecompact.Suite
+	keyID     []byte
+	publicKey []byte
+}
+
+func encodeLV(b []byte) []byte {
+	lengthBuffer := make([]byte, 2)
+	binary.BigEndian.PutUint16(lengthBuffer, uint16(len(b)))
+	return append(lengthBuffer, b...)
+}
+
+func parseODoHTargetConfig(config []byte) (ODoHTarget, error) {
+	if len(config) < 8 {
+		return ODoHTarget{}, fmt.Errorf("Malformed config")
+	}
+	kemID := binary.BigEndian.Uint16(config[0:2])
+	kdfID := binary.BigEndian.Uint16(config[2:4])
+	aeadID := binary.BigEndian.Uint16(config[4:6])
+	publicKeyLength := binary.BigEndian.Uint16(config[6:8])
+	if len(config[8:]) != int(publicKeyLength) {
+		return ODoHTarget{}, fmt.Errorf("Malformed config")
+	}
+
+	suite, err := hpkecompact.NewSuite(hpkecompact.KemID(kemID), hpkecompact.KdfID(kdfID), hpkecompact.AeadID(aeadID))
+	if err != nil {
+		return ODoHTarget{}, err
+	}
+
+	publicKey := config[8:]
+	_, _, err = suite.NewClientContext(publicKey, []byte("odoh query"), nil)
+	if err != nil {
+		return ODoHTarget{}, err
+	}
+
+	// XXX(caw): HPKE doesn't expose Nh so we hard code it here, but maybe it should?
+	keyID, err := suite.Expand(suite.Extract(config, nil), []byte("odoh key id"), 32)
```

Comment:
@jedisct1: flagging here

Reply:
Oh, indeed, the key sizes can be useful. Actually, I don't see any reason not to make everything in the `Suite` structure public. And that was probably the initial intent :) I'll change this shortly.

Reply:
Done!

Reply:
Thanks!